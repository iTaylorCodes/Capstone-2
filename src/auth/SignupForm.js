import { useState } from "react";
import useFields from "../hooks/useFields";
import { useNavigate } from "react-router";
import Alert from "../alert/Alert";

/** Signup Form.
 *
 * Shows a form to signup and manages update to state on changes.
 * On submission, add new user to database and redirects to / route.
 *
 * Routed at /signup
 */
const SignupForm = ({ signup }) => {
  const navigate = useNavigate();
  const [formData, handleChange] = useFields({
    username: "",
    firstName: "",
    lastName: "",
    email: "",
    password: "",
  });
  const [formErrors, setFormErrors] = useState([]);

  const handleSubmit = async (e) => {
    e.preventDefault();
    let result = await signup(formData);
    if (result.success) {
      navigate("/", { replace: true });
    } else {
      setFormErrors(result.errors);
    }
  };

  return (
    <div className="row justify-content-center">
      <div className="col-sm-8 col-md-7 col-lg-6">
        <h1 className="m-3">Create an Account</h1>
        <div className="card">
          <form onSubmit={handleSubmit}>
            <div className="m-3">
              <label htmlFor="username" className="form-label">
                Username
              </label>
              <input
                id="username"
                name="username"
                value={formData.username}
                onChange={handleChange}
                className="form-control"
                placeholder="Username"
                autoComplete="username"
              />
            </div>
            <div className="m-3">
              <label htmlFor="first-name" className="form-label">
                First Name
              </label>
              <input
                id="first-name"
                name="firstName"
                value={formData.firstName}
                onChange={handleChange}
                className="form-control"
                placeholder="First Name"
                autoComplete="firstName"
              />
            </div>
            <div className="m-3">
              <label htmlFor="last-name" className="form-label">
                Last Name
              </label>
              <input
                id="last-name"
                name="lastName"
                value={formData.lastName}
                onChange={handleChange}
                className="form-control"
                placeholder="Last Name"
                autoComplete="lastName"
              />
            </div>
            <div className="m-3">
              <label htmlFor="email" className="form-label">
                Email
              </label>
              <input
                id="email"
                name="email"
                value={formData.email}
                onChange={handleChange}
                className="form-control"
                placeholder="Email"
                autoComplete="email"
              />
            </div>
            <div className="m-3">
              <label htmlFor="password" className="form-label">
                Password
              </label>
              <input
                id="password"
                name="password"
                value={formData.password}
                onChange={handleChange}
                type="password"
                className="form-control"
                placeholder="Password"
                autoComplete="current-password"
              />
            </div>

            {formErrors.length ? (
              <Alert type="danger" messages={formErrors} />
            ) : null}

            <div className="m-3 d-flex justify-content-end">
              <button className="btn btn-primary" type="submit">
                Signup
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  );
};

export default SignupForm;
