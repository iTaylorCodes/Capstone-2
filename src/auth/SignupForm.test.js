import { render, fireEvent, act } from "@testing-library/react";
import SignupForm from "./SignupForm";
import { MemoryRouter } from "react-router";

const signup = jest.fn();

it("renders without crashing", function () {
  render(
    <MemoryRouter>
      <SignupForm signup={signup} />
    </MemoryRouter>
  );
});

it("matches snapshot", function () {
  const { asFragment } = render(
    <MemoryRouter>
      <SignupForm signup={signup} />
    </MemoryRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});

it("shows error alerts", async function () {
  await act(async () => {
    signup.mockReturnValue({ success: false, errors: ["Failed signup"] });
    const { getByText, getByRole } = render(
      <MemoryRouter>
        <SignupForm signup={signup} />
      </MemoryRouter>
    );

    const button = getByText("Signup");
    await fireEvent.click(button);

    const alert = getByRole("alert");
    expect(alert).toBeTruthy();
  });
});

it("changes input values using useFields hook", async function () {
  await act(async () => {
    const { getByLabelText, getByText } = render(
      <MemoryRouter>
        <SignupForm signup={signup} />
      </MemoryRouter>
    );

    const emailInput = getByLabelText("Email");
    await fireEvent.change(emailInput, {
      target: { value: "testtest@test.com" },
    });

    expect(getByLabelText("Email").value).toBe("testtest@test.com");
  });
});

it("reroutes on successful signup", async function () {
  await act(async () => {
    signup.mockReturnValue({ success: true });

    const { getByText } = render(
      <MemoryRouter>
        <SignupForm signup={signup} />
      </MemoryRouter>
    );

    const button = getByText("Signup");
    await fireEvent.click(button);
  });
});
