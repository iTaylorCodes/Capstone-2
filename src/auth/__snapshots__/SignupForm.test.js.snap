// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="row justify-content-center"
  >
    <div
      class="col-sm-8 col-md-7 col-lg-6"
    >
      <h1
        class="m-3"
      >
        Create an Account
      </h1>
      <div
        class="card"
      >
        <form>
          <div
            class="m-3"
          >
            <label
              class="form-label"
              for="username"
            >
              Username
            </label>
            <input
              autocomplete="username"
              class="form-control"
              id="username"
              name="username"
              placeholder="Username"
              value=""
            />
          </div>
          <div
            class="m-3"
          >
            <label
              class="form-label"
              for="first-name"
            >
              First Name
            </label>
            <input
              autocomplete="firstName"
              class="form-control"
              id="first-name"
              name="firstName"
              placeholder="First Name"
              value=""
            />
          </div>
          <div
            class="m-3"
          >
            <label
              class="form-label"
              for="last-name"
            >
              Last Name
            </label>
            <input
              autocomplete="lastName"
              class="form-control"
              id="last-name"
              name="lastName"
              placeholder="Last Name"
              value=""
            />
          </div>
          <div
            class="m-3"
          >
            <label
              class="form-label"
              for="email"
            >
              Email
            </label>
            <input
              autocomplete="email"
              class="form-control"
              id="email"
              name="email"
              placeholder="Email"
              value=""
            />
          </div>
          <div
            class="m-3"
          >
            <label
              class="form-label"
              for="password"
            >
              Password
            </label>
            <input
              autocomplete="current-password"
              class="form-control"
              id="password"
              name="password"
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
          <div
            class="m-3 d-flex justify-content-end"
          >
            <button
              class="btn btn-primary"
              type="submit"
            >
              Signup
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
